helpers.go

```diff
@@ -541,60 +417,60 @@ func getOffer(header string, isAccepted func(spec, offer, specParams string) boo
 
 	// Parse header and get accepted types with their quality and specificity
 	// See: https://www.rfc-editor.org/rfc/rfc9110#name-content-negotiation-fields
-	forEachMediaRange(header, func(accept string) {
+	forEachMediaRange(header, func(accept []byte) {
 		order++
-		spec, quality, params := accept, 1.0, ""
+		spec, quality := accept, 1.0
+
+		var params headerParams
 
-		if i := strings.IndexByte(accept, ';'); i != -1 {
+		if i := bytes.IndexByte(accept, ';'); i != -1 {
 			spec = accept[:i]
 
 			// The vast majority of requests will have only the q parameter with
 			// no whitespace. Check this first to see if we can skip
 			// the more involved parsing.
-			if strings.HasPrefix(accept[i:], ";q=") && strings.IndexByte(accept[i+3:], ';') == -1 {
-				if q, err := fasthttp.ParseUfloat([]byte(strings.TrimRight(accept[i+3:], " "))); err == nil {
+			if bytes.HasPrefix(accept[i:], []byte(";q=")) && bytes.IndexByte(accept[i+3:], ';') == -1 {
+				if q, err := fasthttp.ParseUfloat(bytes.TrimRight(accept[i+3:], " ")); err == nil {
 					quality = q
 				}
 			} else {
-				hasParams := false
-				forEachParameter(accept[i:], func(param, val string) bool {
-					if param == "q" || param == "Q" {
-						if q, err := fasthttp.ParseUfloat([]byte(val)); err == nil {
+				params, _ = headerParamPool.Get().(headerParams) //nolint:errcheck // only contains headerParams
+				fasthttp.VisitHeaderParams(accept[i:], func(key, value []byte) bool {
+					if string(key) == "q" {
+						if q, err := fasthttp.ParseUfloat(value); err == nil {
 							quality = q
 						}
 						return false
 					}
-					hasParams = true
+					params[utils.UnsafeString(utils.ToLowerBytes(key))] = value
 					return true
 				})
-				if hasParams {
-					params = accept[i:]
-				}
 			}
+
 			// Skip this accept type if quality is 0.0
 			// See: https://www.rfc-editor.org/rfc/rfc9110#quality.values
 			if quality == 0.0 {
 				return
 			}
 		}
 
-		spec = strings.TrimRight(spec, " ")
+		spec = bytes.TrimRight(spec, " ")
 
 		// Get specificity
 		var specificity int
 		// check for wildcard this could be a mime */* or a wildcard character *
-		if spec == "*/*" || spec == "*" {
+		if string(spec) == "*/*" || string(spec) == "*" {
```

Comment:
utils.UnsafeString(spec)?

Reply:
Neither go nor gccgo will allocate here.  Not sure about tinygo though :)